platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblem.java

```diff
@@ -86,7 +85,7 @@ public RuntimeException getException() {
     }
 
     @Override
-    public Map<String, Object> getAdditionalData() {
+    public AdditionalData<?> getAdditionalData() {
```

Comment:
🤔 I was expecting something like this:
- Problem gets a type parameter describing what kind of data it uses.
- ProblemDefinition should also have the type parameter, though it probably doesn't need it in the first place
- This getter returns the type parameter.

I understand that you need the type of the builder when adding the additional data. Though as soon as it has been created, you don't need the builder type any more, right? I'll leave a comment to the method in `InternalProblemSpec`.

Reply:
That was essentially my first attempt, but the change got so big that I decided to step back and deal with the removal of the Map type for additionalData. Which is this PR.